pandas/core/ops/array_ops.py

```diff
@@ -132,3 +152,171 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
         result = masked_arith_op(left, right, op)
 
     return missing.dispatch_fill_zeros(op, left, right, result)
+
+
+def arithmetic_op(left, right, op, str_rep, eval_kwargs):
+
+    from pandas.core.ops import (
+        maybe_upcast_for_op,
+        should_extension_dispatch,
+        dispatch_to_extension_op,
+    )
+
+    keep_null_freq = isinstance(
+        right,
+        (
+            ABCDatetimeIndex,
+            ABCDatetimeArray,
+            ABCTimedeltaIndex,
+            ABCTimedeltaArray,
+            Timestamp,
+        ),
+    )
+
+    lvalues = extract_array(left, extract_numpy=True)
+    rvalues = extract_array(right, extract_numpy=True)
+
+    rvalues = maybe_upcast_for_op(rvalues, lvalues.shape)
+
+    if should_extension_dispatch(left, rvalues) or isinstance(
+        rvalues, (ABCTimedeltaArray, ABCDatetimeArray, Timestamp)
+    ):
+        res_values = dispatch_to_extension_op(op, lvalues, rvalues, keep_null_freq)
+
+    else:
+        with np.errstate(all="ignore"):
+            res_values = na_arithmetic_op(lvalues, rvalues, op, str_rep, eval_kwargs)
+
+    return res_values
+
+
+def comparison_op(left, right, op):
+    from pandas.core.ops import should_extension_dispatch, dispatch_to_extension_op
+
+    # NB: We assume extract_array has already been called on left and right
```

Comment:
adding type annotations would help here? or do we allow scalar `left` and `right`, which makes this difficult?

Reply:
I see in the commits you reverted the added types again (because mypy was not happy), but can you then leave them as comments? As for the human reader they can still be very valuable? 
(Or add the equivalent information as a docstring)

Reply:
The particular mypy complaint has come up before (in indexes.base).  I'm hoping to get a PR up this morning that addresses it, so I can add non-comment types here in a follow-up

Reply:
Was the error you were referring to `pandas/core/ops/array_ops.py:333: error: Cannot call function of unknown type`? That's what I was seeing in the CI log

If so I think this is a relevant issue on mypy:

https://github.com/python/mypy/issues/2248

Maybe using an actual function for `fill_int` instead of a lambda resolves

Reply:
The complaint was about the line in `Index.__new__` calling `cls._scalar_data_error(data)`.  I think that can be fixed by changing it to `raise cls._scalar_data_error(data)`

Reply:
is there a preferred way of typing e.g. "one of these N objects", in particular the `op` arg in these functions.  Could use an enum?

Reply:
And I think I asked this already, but double checking: we dont have a good way of typing "Any but not a DataFrame"?

Reply:
> is there a preferred way of typing e.g. "one of these N objects", in particular the `op` arg in these functions. Could use an enum?

Is it a `Union` you are looking for?



> And I think I asked this already, but double checking: we dont have a good way of typing "Any but not a DataFrame"?

Not that I am aware of - not sure that would be possible to determine that statically

Reply:
> Is it a Union you are looking for?

I'm thinking of the `op` arg which is going to be among `{operator.and_, operator.or_, operator.xor, roperator.rand_, roperator.ror_, roperator.rxor}`.

Reply:
Ah ok. No Literal types are still an experimental thing - not available in stdlib so best for now is just Callable